### DISCUSSION TITLE
openpgp.min.js and JavaScriptCore iOS

### DISCUSSION THREAD

**Discussion by boby1975** [openpgpjs/openpgpjs#1213] [Q&A]
Hi guys,
And I'm trying to use openpgp.min.js from dist folder with JavaScriptCore in iOS app (swift).
But there is such error when trying to open this file under VM JavaScriptCore.
"TypeError: undefined is not a constructor (evaluating \'new t.Set([a.default.hash.md5,a.default.hash.ripemd])\')"

Does anybody know why it happens? And how it can be fixed?
Thanks a lot

**Comment by twiss** [comment]
Hey :wave: Probably this means that [Sets](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) are not supported.
What you could do is locally clone this repository, revert https://github.com/openpgpjs/openpgpjs/commit/15202d9d400419d34a4e7b32e9ca7ba1ba85a4b2, run `npm run build`, and then use the `dist/compat/openpgp.min.js` build.
However - you'll likely run into many more things that are missing or not supported in plain JavaScriptCore - such as anything related to crypto, and getting randomness.
So you would have to reimplement those, as well, by getting randomness from the host OS, for example.
You might have more luck trying to run it in [nodejs-mobile](https://github.com/JaneaSystems/nodejs-mobile) or [LiquidCore](https://github.com/LiquidPlayer/LiquidCore) - though some others were trying to use the latter, and also had some issues, e.g. see #1080. But there seems to be some effort ongoing to get that working.

**Comment by boby1975** [comment]
Thank you. You're right. With your help one error went away but a new one appeared: "ReferenceError: Can\'t find variable: navigator". 
Really no one was able to use openpgp.js in iOS without a WebView or similar? 
The main task is "streaming encrypt and decrypt" of large files using openpgp.js but I still can’t find a proper working solution (((

**Comment by twiss** [comment]
It's probably possible, but for streaming encryption and decryption, you probably also want to have access to disk and/or the network in JS, so that you don't have to copy each chunk to the host language while streaming? So I would again recommend looking into the mobile Node.js solutions.

**Comment by boby1975** [comment]
Yes. This is complex task and with Node.js it’s probably easier to do solution. But still how I can build openpgp.js with crypto module inside because VM JavaScriptCore does not have a built-in like Node.js or web-browser?  
Thanks in advance.

**Comment by twiss** [comment]
For the example I gave (getting cryptographically random bytes) this is likely not possible without some type of communication with the host. There is no way for a module in JS to come up with secure random bytes by itself. So you will have to polyfill `crypto.getRandomValues`, for example, getting random bytes from the host OS.
For the rest, OpenPGP.js doesn't actually need Node.js's crypto module, most things are also implemented in JS. It will be slower than the native Node.js crypto implementation, but polyfilling Node.js's crypto module inside the build won't necessarily help with that, so 🤷‍♂️

**Comment by tomholub** [comment]
We use JavaScriptCore on iOS happily. It works, but you have to do quite a bit of the low-level crypto yourself, because else it either wouldn't work or be painfully slow.
